### PR TITLE
TRUNK-6058 : Create CI Process to create liquibase snapshots

### PR DESCRIPTION
**Pull request by ManojLL**
## Description of what I changed

I write a script to automate the liquibase snapshot generating process
## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-6058

## Checklist: I completed these to help reviewers :)

- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by ManojLL** [comment]
@dkayiwa @ibacher could you please review this work?

**Comment by dkayiwa** [comment]
@ManojLL as per this https://wiki.openmrs.org/display/docs/Pull+Request+Tips you need to transition the JIRA ticket state to indicate that you are waiting for review.

**Comment by dkayiwa** [comment]
Are you able to run this on our bamboo CI server?

**Comment by ManojLL** [comment]
> Are you able to run this on our bamboo CI server?

This has not integrate with CI process yet, I just need to review my work and then we can run this in CI process.

**Comment by ManojLL** [comment]
@dkayiwa here are the steps to generate liquilbase snapshot automatically

1. set ENV variable called `OPENMRS_INSTALLATION_SCRIPT`
 `export OPENMRS_INSTALLATION_SCRIPT=<path to openmrs-core folder>/liquibase/scripts/installation.properties`
2. run the script, `username` and `passowrd` refer to a MySQL user:
`cd <path to openmrs-core folder>`
` . liquibase/scripts/generate_liquibase_snapshots.sh <username> <password>`

**Comment by dkayiwa** [comment]
Don't i need to set some sort of permissions to make the script an executable?
Can we also avoid running the tests?

File: liquibase/scripts/installation.properties [hunk 1]
```diff
@@ -0,0 +1,17 @@
+install_method=auto
+connection.url=jdbc:mysql://localhost:3306/@DBNAME@?autoReconnect=true&sessionVariables=default_storage_engine=InnoDB&useUnicode=true&characterEncoding=UTF-8
+connection.driver_class=com.mysql.cj.jdbc.Driver
+connection.username=root
```
Comment (dkayiwa) [inline liquibase/scripts/installation.properties:4]: As per the instructions you gave, i do not see how the username and password is passed over.
No wonder i got `Access denied for user 'root'@'localhost' (using password: YES)` when i tried to run this in spite of the fact that i supplied a correct username and password.

File: liquibase/scripts/generate_liquibase_snapshots.sh [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    mysql -u "${1}" -p"${2}" < "liquibase/scripts/drop_openmrs_schema.sql"
+    mysql -u "${1}" -p"${2}" < "liquibase/scripts/create_openmrs_database.sql"
+}
+
+function run_openmrs_core_application() {
+ # Change directory to the specified root folder
+	# Change directory to the 'webapp' folder
+	cd webapp
+	
+	# Remove the ~/.OpenMRS directory
```
[the remark this answers is not in the document]
Comment (ManojLL) [reply liquibase/scripts/generate_liquibase_snapshots.sh:23]: This is what I was thinking about,
 I just trying to automate the snapshot generating steps mentioned [here](https://github.com/openmrs/openmrs-core/tree/master/liquibase#generating-and-applying-liquibase-snapshots). However, we can not drop existing can't delete our current production-level database and application while doing this. so alter this scrip to generate the snapshots without dropping current database and existing application data and this worked successfully and generate liquibase snapshots without any issue. 

 I'm not quite sure about the generated snapshots are the same when using these two methods.

**Comment by ManojLL** [comment]
>
Can we also avoid running the tests? yes we can , when make a release we run the tests so I think we do not need to run the test again in the script.

**Comment by dkayiwa** [comment]
So can you make the necessary changes?

**Comment by ManojLL** [comment]
> So can you make the necessary changes?

Update the script to generate snapshot without dropping the existing database.

**Comment by dkayiwa** [comment]
@ManojLL is this still Draft?

File: liquibase/scripts/generate_liquibase_snapshots.sh [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    cd liquibase/
+    . scripts/test_liquibase_snapshots.sh "${1}" "${2}"
+    cd ..
+}
+function echo_usage() {
+	echo "usage: . liquibase/scripts/create_liquibase_snapshots.sh <username> <password>"
+}
+
+project_version=$(grep -m 1 '<version>' pom.xml | sed -n 's/.*<version>\(.*\)<\/version>.*/\1/p') 
+openMRS_version=$(echo $project_version | sed "s/\([0-9]*\.[0-9]*\)\..*/\1/")
```
Comment (dkayiwa) [inline liquibase/scripts/generate_liquibase_snapshots.sh:74]: Is there a reason why for some you are using capital letters and then others lower case? `openMRS_version` vs `new_openmrs_version`

File: liquibase/scripts/installation.properties (hunk 1 above)
Comment (dkayiwa) [reply liquibase/scripts/installation.properties]: > I just realized I missed mentioning a couple of instructions earlier, my bad!

So where are you putting the instructions?

File: liquibase/scripts/generate_liquibase_snapshots.sh [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        sed -i '/private static final List<String> SNAPSHOT_VERSIONS = Arrays.asList(/s/);/,\ "'"${openMRS_version}"'"\);/' "$change_log_java_file"
+    fi
+    
+    if ! grep -q "${new_openmrs_version}" "$change_log_java_file"; then
+        sed -i '/private static final List<String> UPDATE_VERSIONS = Arrays.asList(/s/);/,\ "'"${new_openmrs_version}"'"\);/' "$change_log_java_file"
+    fi
+    
+    # Update the version in openmrs_version.txt only if it's a new version
+    if [ "${openMRS_version}" != "${new_openmrs_version}" ]; then
+        sed -i "s/\(openmrs_version.txt=\).*/\1${new_openmrs_version}.0-SNAPSHOT/" "liquibase/scripts/fix_liquibase_snapshots.sh"
```
Comment (dkayiwa) [inline liquibase/scripts/generate_liquibase_snapshots.sh:61]: Did you really test the above line?

Comment (ManojLL) [reply liquibase/scripts/generate_liquibase_snapshots.sh:61]: This is not required now, missed to remove this line from the script when ,I will remove these lines

File: liquibase/scripts/generate_liquibase_snapshots.sh (hunk 3 above)
Comment (ManojLL) [reply liquibase/scripts/generate_liquibase_snapshots.sh:74]: No reason to name like this

File: liquibase/scripts/generate_liquibase_snapshots.sh [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    mv liquibase/snapshots/liquibase-core-data-UPDATED-SNAPSHOT.xml api/src/main/resources/org/openmrs/liquibase/snapshots/core-data/"liquibase-core-data-$openmrs_version.x.xml"
+    mv liquibase/snapshots/liquibase-schema-only-UPDATED-SNAPSHOT.xml api/src/main/resources/org/openmrs/liquibase/snapshots/schema-only/"liquibase-schema-only-$openmrs_version.x.xml"
+}
+
+function create_new_liquibase_update_file() {
+    new_liquibase_file="liquibase-update-to-latest-${new_openmrs_version}.x.xml"
+    cp "api/src/main/resources/liquibase-update-to-latest-template.xml" "api/src/main/resources/org/openmrs/liquibase/updates/$new_liquibase_file"
+    
+    database_updater_class="api/src/test/java/org/openmrs/util/DatabaseUpdaterDatabaseIT.java"
+    variable_name="CHANGE_SET_COUNT_FOR_GREATER_THAN_2_1_X"
```
Comment (dkayiwa) [inline liquibase/scripts/generate_liquibase_snapshots.sh:30]: Why are we explicitly using 2.1.x here?

File: liquibase/scripts/generate_liquibase_snapshots.sh [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    
+    # Increase the variable value by 1
+    sed -i "s/\($variable_name *= *\)[0-9]\+/\1$(( $(grep -o "$variable_name *= *[0-9]\+" "$database_updater_class" | awk '{print $NF}') + 1 ))/" "$database_updater_class"
+    
+    insert_line="	<include file=\"org/openmrs/liquibase/updates/$new_liquibase_file\"/>"
+    sed -i -e '$!N;$!N;$i\'"$insert_line" "api/src/main/resources/liquibase-update-to-latest-from-1.9.x.xml"
```
Comment (dkayiwa) [inline liquibase/scripts/generate_liquibase_snapshots.sh:36]: Why are we using 1.9.x here?

File: liquibase/scripts/installation.properties (hunk 1 above)
Comment (dkayiwa) [reply liquibase/scripts/installation.properties]: @ManojLL did you see the above question?

Comment (ManojLL) [reply liquibase/scripts/installation.properties]: [this is where ](https://github.com/openmrs/openmrs-core/pull/4474#issuecomment-1844327017) I put the instructions for run the script. what I forgot to mention is `chmod +x <path to script>`.

File: liquibase/scripts/generate_liquibase_snapshots.sh (hunk 5 above)
Comment (ManojLL) [reply liquibase/scripts/generate_liquibase_snapshots.sh:30]: according to the steps mentioned [here](https://github.com/openmrs/openmrs-core/tree/master/liquibase#step-1---add-the-new-snapshot-files), we are copying the liquibase-updates template and rename it to current version. That template file contains change set as below (this include a comment), 
 `<changeSet id="42ddb873-f064-4418-a6c0-6c09426c4832" author="wschlegel">
		<comment>Empty change set for integration tests.</comment>
	</changeSet>` 

So when adding a new change-set we need to increase the `CHANGE_SET_COUNT_FOR_GREATER_THAN_2_1_X` value by one.

File: liquibase/scripts/generate_liquibase_snapshots.sh (hunk 6 above)
Comment (ManojLL) [reply liquibase/scripts/generate_liquibase_snapshots.sh:36]: if the liquibase snapshot update version is above 1.9.x , we add that names in this file. This also mentioned in [here](https://github.com/openmrs/openmrs-core/tree/master/liquibase#step-2---create-a-new-liquibase-update-file)

File: liquibase/scripts/installation.properties (hunk 1 above)
Comment (dkayiwa) [reply liquibase/scripts/installation.properties]: Please do not keep instructions in a github comment.

File: liquibase/scripts/generate_liquibase_snapshots.sh (hunk 5 above)
Comment (dkayiwa) [reply liquibase/scripts/generate_liquibase_snapshots.sh:34]: But the current version is not 2.1.?

File: liquibase/scripts/generate_liquibase_snapshots.sh (hunk 6 above)
Comment (dkayiwa) [reply liquibase/scripts/generate_liquibase_snapshots.sh:40]: In the documentation that you referenced above, isn't 1.9.x just an example. Where the actual value depends on the current version?

File: liquibase/scripts/generate_liquibase_snapshots.sh (hunk 5 above)
Comment (ManojLL) [reply liquibase/scripts/generate_liquibase_snapshots.sh:34]: when adding new `changeset` in a new version which is greater than 2.1, we should increase this `CHANGE_SET_COUNT_FOR_GREATER_THAN_2_1_X` value by the number of newly added changesets.

File: liquibase/scripts/generate_liquibase_snapshots.sh (hunk 6 above)
Comment (ManojLL) [reply liquibase/scripts/generate_liquibase_snapshots.sh:40]: The file called liquibase-update-to-latest-from-1.9.x.xml is used for integration tests. When we make a new snapshot, we create a file for the next snapshot version using the format liquibase-update-<snapshot-next-version>.x.xml. After creating this new file, we need to add its path to the liquibase-update-to-latest-from-1.9.x.xml file.

File: liquibase/scripts/installation.properties (hunk 1 above)
Comment (dkayiwa) [reply liquibase/scripts/installation.properties]: Did you address the above comment?

Comment (ManojLL) [reply liquibase/scripts/installation.properties]: @dkayiwa Do I need to update liquibase read me file with the instructions ?

Comment (dkayiwa) [reply liquibase/scripts/installation.properties]: Yes. 😊

Comment (ManojLL) [reply liquibase/scripts/installation.properties]: @dkayiwa I added the instructions to the `Readme.md` file.

File: liquibase/README.md [hunk 7]
```diff
@@ -125,7 +125,16 @@ need to be updated so that they include the change.
...
 database and applying snapshots to an OpenMRS database.
 
-### How to generate Liquibase snapshots
+### How to generate Liquibase snapshots automatically
+#### step 1 - checkout to project root folder
+ 	cd <some root folder>/openmrs-core
+#### step 2 - run the script `openmrs/liquibase/scripts/generate_liquilbase_snapshots.sh`:
+Run the following commands to generate the Liquibase snapshots automatically, where username and password refer to a MySQL user:
+
+	. liquibase/scripts/generate_liquibase_snapshots.sh <username> <password>
```
Comment (dkayiwa) [inline liquibase/README.md:134]: After running this script, can you list the files which will get created and their output locations?

**Pull request closed without merging**